### PR TITLE
Add alias feature for LG import

### LINKED ISSUE #5341: Allow package prefixes in LG imports and handle templates naming conflicts
This is a sub-ticket to #5298 which contains everything needed to enable package prefix on resource level.

This sub-ticket is to track the effort on LG side, 
* support the package prefix in LG statement, like `[import](package.name#lg.name)` 
* handle naming conflicts of LG templates, by giving an name\alias when importing, like `[import](package.name#lg.name) as foo`, this will allow calling the conflicted template inside the imported LG file, like foo.Welcome()

detailed design is in the original ticket and doc.

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #5341

## Description
This PR aims to enable LG alias feature.
What is LG alias feature? 

### Current situation
In the exist import case, the templates of import cannot conflict with the existing template names. 
For example:
b.lg:
``` 
[import](a.lg)
# template1
- hi
```

a.lg:
``` 
# template1
- hello
```
An error would be thrown, because the template name "template1" is used both in `b.lg` and the base lg `a.lg`.

### New concept 
Originally, there only exists one import mode, that is, all the templates in the imported lg are introduced into parent's LG.
This PR, we introduce a concept, called "import alias", which is already in some programming languages. Like:
```
C# :  using alias = Some.NameSpace;
js ES6: import { export1 as alias1 } from "module-name";
```
We also introduce a similar format, which is compatible with the original format. Here is the detail and the comparison with other languages.

LG/class layer:

Language | import target (class, lg) | import target(class, lg) with alias
-- | -- | --
C# | using static NameSpace.ClassName | Not support
js es6 | import {className} from 'module-name' | import { className as alias }  from "module-name";
lg | \[description](namespace#a.lg) | \[description](namespace#a.lg) as lgalias

Namespace layer:(In the future )
Language | import namespace | import namespace with alias
-- | -- | --
C# | using NameSpace | using alias = NameSpace
js es6 | import * from 'module-name' | import * as name from "module-name";
lg *(in design) | \[description](namespace#)  | \[description](namespace#) as namespaceAlias

### How to use
- If you want to import all templates into current LG file, and handler the naming conflict, keep the original format `[import](xx.lg)` is enough.
- If you want to import a lg file (same or different namespace), and want to assign an alias to it, use `[import](xx.lg) as alias` works for you.

For import with alias, the principles:
- Allow duplicate template names
- If user call the duplicated name directly, the template in current lg file would be evaluated firstly
- Use the alias as the prefix if user wants to call the imported templates
- If the template name is duplicated with built-in function, call built-in function firstly

### Examples
Here are some examples to show how to evaluate the corresponding template in the right way.
Base1.lg
```
# mytemplate
- my template in base1

# length
- length in base1

# template
- template owns by base1
```

Base2.lg
```
# mytemplate
- my template in base2

# length
- length in base2

# Base2Template
- template owns by base2
```

main.lg
```
[import](Base1.lg) as base1
[import](Base2.lg)

> invalid!  Base2Template is conflict with the name in "Base2.lg"
> # Base2Template
> - hi

# template
- template in main
```
Here are the evaluate results:

template name | result | remark
-- | -- | --
mytemplate | my template in base2 | call base2's template
base1.mytemplate | my template in base1 | call base1's template
length | number | call built-in function's length
lg.length | length in base2 | call base2's template
base1.length | length in base1 | call base1's template
template | template in main | call main's template
base1.template | template owns by base1 | call base1's template

## Specific Changes
1. Add alias support for LG
2. Refactor the constructor of Analyzer/Evaluator/Expander/StaticChecker
3. Split "#" in default File import resolver

## Tests
- Alias unit tests
- Error alias unit tests
- Dialog E2E tests

### CODE REVIEW

**Review by boydc2014 — DISMISSED** [review]

**Comment by boydc2014** [comment]
@chrimc62 do you want take a look? LGTM

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateImport.cs [hunk 1]
```diff
@@ -47,10 +49,24 @@ internal TemplateImport(string description, string id, SourceRange sourceRange)
         /// </value>
         public SourceRange SourceRange { get; set; }
 
+        /// <summary>
+        /// Gets or sets the "as" syntax. For example: [import](path) as myAlias.
+        /// </summary>
+        /// <value>
+        /// Imported with the "as" syntax. For example: [import](path) as myAlias.
+        /// </value>
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateImport.cs:57]: Alias for template.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 2]
```diff
@@ -45,6 +45,7 @@ public class Templates : List<Template>
         /// <param name="importResolver">Resolver to resolve LG import id to template text.</param>
         /// <param name="options">List of strings representing the options during evaluating the templates.</param>
         /// <param name="source">Templates source.</param>
+        /// <param name="namedReferences">References that imported with the "as" syntax，for example: [import](path.lg) as myAlias.</param>
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:48]: >References that imported with the "as" syntax，for example: [import](path.lg) as myAlias. [](start = 42, length = 88)

Map from import alias to templates.

**Review by chrimc62 — APPROVED** [review]
:shipit:

**Comment by chrimc62** [comment]
This looks awesome!  Thanks for doing the work!

---
In reply to: [806364128](https://github.com/microsoft/botbuilder-dotnet/pull/5342#issuecomment-806364128) [](ancestors = 806364128)

**Pull request merged**